### DISCUSSION TITLE
Использование подготовленных кем-то IP-листов - плохая идея

### DISCUSSION THREAD

**Discussion by Ground-Zerro** [Ground-Zerro/DomainMapper#50] [General]
Сказанное ниже относится как к готовым IP-листам, так и к файлам с готовыми маршрутами.

**CDN**
Многие сервисы и веб-хостинги на которых размещены сайты используют CDN - сеть доставки контента. Таким наиболее распространенным хостером является [Cloudflare](https://www.cloudflare.com/). Некоторые сервисы используют свои собственные CDN.
В случае использования сайтом/сервисом CDN его IP-адрес будет отличаться для разных устройств (хостов) в зависимости от их геопозиции, текущей нагрузки на сервера, времени суток и еще ряда факторов.
Кроме того, IP-адреса для одного и того же списка доменов, сформированные на одном и том же хосте в разное время, даже спустя всего минуту, так же будут различаться.

**Использование разных DNS серверов**
Вы не знаете какой DNS сервер использовал тот, кто сделал IP-лист (маршруты).
Для условного `site.com` GoogleDNS вернёт `100.20.20.10`, Quad9DNS вернёт `30.44.33.77`, CloudflareDNS вернет `66.77.22.45` и т.д. *(Это пример, не нужно воспринимать его буквально!)*
Для `site.com` составителю вернулся IP-адрес от GoogleDNS, а когда `site.com` открывает другой человек, с установленным провайдерским или другим публичным DNS, ему возвращается совершенно другой IP-адрес, маршрута к которому нет.
Бывают ситуации, когда на хосте указано несколько DNS, причем от разных сервисов. В этом случае при открытии `site.com` ответ может вернуться от CloudflareDNS с IP-адресом которого нет в маршрутах, а при повторном обращении к `site.com` от GoogleDNS с другим «правильным» и тогда все **Ок**, а может и от Quad9DNS и тогда снова мимо.

Получается такая ситуация: Маршруты добавлены для одного IP, а DNS вернул другой, часть трафика пошла через туннель, часть напрямую, что-то работает, а что-то нет.

**Пример для "Antifilter community edition":**
Первый список доменов обработан на локальной машине в России, второй на VPS в Нидерландах:
[List1.txt](https://github.com/user-attachments/files/17448058/List1.txt)
[List2.txt](https://github.com/user-attachments/files/17448057/List2.txt)
Результаты сравнения двух списков IP-адресов:
В первом списке есть 1099 IP-адресов, которых нет во втором.
Во втором списке есть 993 IP-адреса, которых нет в первом.
1276 IP-адресов присутствуют в обоих списках.
Различие между двумя списками составляет около **63%**. 
***Списки отличаются более чем на половину!!***
Т.е. 63% или больше половины IP-адресов, полученных на одной машине окажутся полностью бесполезны на другой.  Они будут просто висеть в таблице маршрутизации мертвым грузом, увеличивая время ее обработки роутером и задержки маршрутизации (незначительно, но все же).

**Итого:**
Резолвить домены в IP-адреса на одной машине с целью использования этого списка на другой не всегда приведет к желаемому результату - не на всех хостах будет работать полноценно проще говоря.
Да это хоть как-то работает и вполне устраивает большинство, но такой подход в купе с его простотой несет в себе и множество минусов.

**Варианты решения проблемы:**
При наличии роутера с Linux ядром лучше использовать раздельную маршрутизацию на основе доменных списков, разделяя их трафик между тоннелями. Главное преимущество такой схемы в том, что домен резолвится в IP-адрес на конкретном устройстве и удаленный хост 100% будет ему (IP-адресу) соответствовать и обмениваться трафиком через указанного пользователем провайдера или VPN тоннель.

Есть готовые решения для Роутеров на OpenWRT, KeeneticOS, RouterOS и других.

**Еще вариант:**
Если роутер простой, но к нему можно подключиться по SSH и у вас арендован VPS можно попробовать [DNS-RouteSync-Navigator](https://github.com/Ground-Zerro/DNS-RouteSync-Navigator). Но это опять же компромисс, соответствующий ситуации.

**Comment by Sarabanga** [comment]
Автор видимо сам не понял что хотел сказать.

> Да это хоть как-то работает и вполне устраивает большинство, но я считаю такой способ неполноценным.

А как иначе я должен достучаться до домена с CDN?
Я должен упорно запрашивать местный заблокированный адрес и если его окончательно не заблокировали то я должен прописать на него маршрут на роутере через VPN и ходить на местный сервер через Нидерланды? Смысл? Неукротимое желание высокого пинга?
А выход как раз один. Я должен получить ближайший адрес от моего VPS в Нидерландах и настроить на него маршрутизацию на роутере.
Другого варианта нет для работы с тем же Cloudflare.

**Comment by Ground-Zerro** [comment]
@Sarabanga 
Не уверен что я правильно понял комментарий, но попробую ответить.

> я должен прописать на него маршрут на роутере через VPN и ходить на местный сервер через Нидерланды?

Так и получится если зарезолвить домены c домашней машины и кинуть их в статические маршруты перенаправив к ним трафик через "Нидерланды".
Некоторые резолвят домены на VPS в "Нидерландах", а в роутере оставляют DNS сервером тот, что выдал домашний провайдер. Или прописывают статические маршруты, сделанные неизвестно кем и неизвестно где. И еще множество ситуаций.
Пост какраз об этом.

Совершенно правильно отмечено, что резолвить домены нужно понимая, что ты делаешь.

IP условного `site.com`, хостящегося у CDN-провайдера не постоянен даже для одного и того же устройства и может меняться. Сейчас один, через минуту или день уже другой.

> получить ближайший адрес от моего VPS

К слову CDN это не только геораспределенная сеть и трафик не всегда идет именно к **ближайшему** для хоста узлу. Есть еще факторы влияющие на возвращаемый IP-адрес DNS имени к которому CDN отправит клиента в т.ч. например нагрузка на сервер в данный момент времени , настройки безопасности и другие.

PS
Есть что-то не получается напиши немного подробней и не так эмоционально, я постараюсь помочь.

**Comment by rekryt** [comment]
Посмотрите в сторону реализации обхода через ip-зоны:
https://github.com/rekryt/iplist - инструмент который вы можете самостоятельно себе захостить, собирать ip  адреса нужных сайтов и их ip-зоны через whios (из ripn) автоматически с заданной периодичностью обновления.

https://iplist.opencck.org/
